sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Serialization/SchemaRegistrySerializer.cs

```diff
@@ -0,0 +1,550 @@
+            else if (_serializerOptions.Format == SchemaFormat.Json)
+            {
+                _mimeType = JsonMimeType;
+            }
+            else
+            {
+                _mimeType = CustomMimeType;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SchemaRegistrySerializer"/> class for mocking use in testing.
+        /// </summary>
+        /// <remarks>
+        /// This constructor exists only to support mocking. When used, class state is not fully initialized, and
+        /// will not function correctly; virtual members are meant to be mocked.
+        ///</remarks>
+        protected SchemaRegistrySerializer()
+        {
+        }
+
+        private readonly LruCache<string, string> _idToSchemaMap = new(CacheCapacity);
+        private readonly LruCache<string, string> _schemaToIdMap = new(CacheCapacity);
+
+        #region Serialize
+        /// <summary>
+        /// Serializes the message data and stores it in <see cref="MessageContent.Data"/>. The <see cref="MessageContent.ContentType"/>
+        /// will be set to "application/json+schemaId" if serializing with JSON, "avro/binary" if serializing with Avro, and "text/plain+schemaId"
+        /// otherwise, where schemaId is the Id of the schema that was generated from the type.
+        /// </summary>
+        /// <param name="data">The data to serialize into the message.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        /// <typeparam name="TMessage">The <see cref="MessageContent"/> type to hold the serialized data.</typeparam>
+        /// <typeparam name="TData">The type of the data being serialized.</typeparam>
+        /// <exception cref="InvalidOperationException">
+        ///   This can occur if the <c>groupName</c> was not specified when constructing the <see cref="SchemaRegistrySerializer"/>.
+        ///   It can also occur if the <typeparamref name="TMessage"/> type does not have a public parameterless constructor.
+        /// </exception>
+        /// <exception cref="RequestFailedException">
+        ///   An error occurred while attempting to communicate with the Schema Registry service. This can occur if the schema generated from the type
+        ///   does not exist in the registry.
+        /// </exception>
+        /// <exception cref="Exception">
+        ///   The data did not adhere to the specified schema, or the schema itself was invalid. The inner exception will hold more detailed information about
+        ///   the exception.
+        /// </exception>
+        public TMessage Serialize<TMessage, TData>(
+            TData data,
+            CancellationToken cancellationToken = default) where TMessage : MessageContent, new()
+            => (TMessage)SerializeInternalAsync(data, typeof(TData), typeof(TMessage), false, cancellationToken).EnsureCompleted();
```

Comment:
Why is this doing sync over async given the internally used `ObjectSerializer` provides both a synchronous and an asynchronous path? I mean I see the async flag but that seems to be still a risky thing considering the async method has to properly check it everywhere to avoid any yielding which then leads to sync over async.

Reply:
This is a common pattern that we use to avoid a lot of duplicated code. We have a private/internal method that is called for both sync and async code paths and pass the `async` flag based on where it is being called from. So this isn't actually sync over async because for the sync path the task will be completed.